### DISCUSSION TITLE
Stop making window flicker with draw()

### DISCUSSION THREAD

**Discussion by tmiethlinger** [alandefreitas/matplotplusplus#237] [Q&A]
Hi,
I really like the library and I would like to use it for some sort of simple animation.

The following code works, but unfortunately the window "flickers" whenever the window updates / I call draw. I hope it is clear, what I mean. Is it possible to somehow fix that?

#include <cmath>
#include <matplot/matplot.h>

#include <unistd.h>
inline void mysleep(unsigned millis)
{
    usleep(millis * 1000);
}

int main()
{
    using namespace matplot;
    auto f = figure(true);
    auto ax = gca();
    std::vector<double> x = linspace(0, 2*pi);
    axis({0, 2*pi, -1, 1});
    std::vector<double> y = transform(x, [](auto x) { return sin(x); });

    plot(ax, x, y);
    f->draw();

    double theta = 0;
    while(true)
    {
        y = transform(x, [theta](auto x) { return sin(x + theta); });
        plot(ax, x, y);
        f->draw();
        mysleep(100);
        theta += 0.01;
    }

   return 0;
}

Thanks a lot!

Thomas

**Comment by alandefreitas** [comment]
I don't it's possible to change that with gnuplot + pipes. It's not really meant for high FPS. This will probably only be fixed in matplot++ v2.

**Reply by tmiethlinger** [reply]
Ok, thank you for your response! The thing is: I originally found the repo [https://github.com/dstahlke/gnuplot-iostream](https://github.com/dstahlke/gnuplot-iostream) where they have an animation example ("demo_animation" in example-misc.cc). There, I can increase the FPS very high (here: 10 milliseconds) without problem. Do you know by any chance the reason for the different behavior? Do they not use pipes (since you wrote you use gnuplot + pipes)?

By the way, I find that gnuplot-iostream is kinda difficult to use. I was able to animate one function, but I didn't understand the syntax for animating two functions... which is why I tried matplot++.
I.e., I tried the following (and different variants thereof)

#include <fstream>
#include <vector>
#include <map>
#include <limits>
#include <cmath>
#include <cstdio>

#include "gnuplot-iostream.h"

#define M_PI 3.14159265358979323846

#include <unistd.h>
inline void mysleep(unsigned millis)
{
    ::usleep(millis * 1000);
}

void animation()
{
    Gnuplot gp;

    std::cout << "Press Ctrl-C to quit (closing gnuplot window doesn't quit)." << std::endl;

    gp << "set yrange [-1:1]\n";

    const int N = 1000;
    std::vector<double> f1(N);
    std::vector<double> f2(N);

    double theta = 0;
    while(true)
    {
        for(int i=0; i<N; i++)
        {
            double alpha = (static_cast<double>(i)/N - 0.5) * 10;
            f1[i] = sin(alpha*8.0 + theta) * exp(-alpha*alpha/2.0);
            f2[i] = sin(alpha*8.0 + theta) * exp(-alpha*alpha/1.0);
        }

        // 1 function: works
        gp << "plot" << " '-' binary" << gp.binFmt1d(f1, "array") << "with lines title 'f1(x)'\n";
        gp.sendBinary1d(f1);
        gp.flush();

        // 2 functions: doesn't work
        //gp << "plot" << " '-' binary" << gp.binFmt1d(f1, "array") << "with lines notitle";
        //gp.sendBinary1d(f1);
        //gp << "plot" << " '-' binary" << gp.binFmt1d(f2, "array") << "with lines notitle\n";
        //gp.sendBinary1d(f2);
        //gp.flush();

        mysleep(10);
        theta += 0.05;
    }
}

int main(int argc, char *argv[])
{
    animation();
}

I thought that you might find that interesting :-) I added a video just for reference.
Anyway, thank you for your answer and the great library!

Best regards, Thomas

https://user-images.githubusercontent.com/91594368/154856772-f86888b3-69ad-4ac9-a0d5-3a47306ad707.mp4
